blog/_src/posts/2017-02-05-on-application.md

```diff
@@ -0,0 +1,108 @@
+`if`[racket] from `racket/base`[racket], rather it refers to whatever the enclosing environment
+defines `#%app`[racket] to be (which by default means ordinary function application from
+`racket/base`[racket]).
+
+However, imported modules can provide their own definitions of function application by providing an
+`#%app`[racket] macro. Let's define our own `#%app`[racket] that, in addition to applying a
+function, prints out a trace message. First let's define a helper function to implement the tracing:
+
+<!-- more -->
+
+```racket
+#lang racket
+(define (trace f . args)
+  (printf "Applying ~v to arguments ~v\n" f args)
+  (apply f args))
+```
+
+We use `~v`[racket] to more clearly see the difference between strings, symbols, and numbers. Now,
+we can write a macro to insert a call to our function wherever we use the macro:
+
+```racket
+#lang racket
+(define-syntax-rule (trace-app f arg ...)
+  (trace f arg ...))
+```
+
+This macro is very simple, and on it's own isn't really useful at all (we could just call `trace`
+directly). However, we can provide this macro with the name `#%app`[racket] to trigger the automatic
+use of `trace-app` whenever a function call is written. We'll move our `trace` and `trace-app`
+definitions into a submodule to see this in action without multiple files:
+
+```racket
+#lang racket
+(module trace racket
+  (define (trace f . args)
+    (printf "Applying ~v to arguments ~v\n" f args)
+    (apply f args))
+  (define-syntax-rule (trace-app f arg ...)
+    (trace f arg ...))
+  (provide (rename-out [trace-app #%app])))
+```
+
+Now, with a simple `require`[racket] statement and these seven lines of code, we can trace the order
+of evaluation of all expressions in a Racket module.
+
+```racket
+> (require 'trace)
+
+> (+ (string-length "racket")
+     (string-length "application"))
+Applying #<procedure:string-length> to arguments '("racket")
+Applying #<procedure:string-length> to arguments '("application")
+Applying #<procedure:+> to arguments '(6 11)
+17
+```
+
+## The practice
+
+Redefinition of `#%app`[racket] is occasionally used when defining new languages, but a much more
+pedestrian use is to add some notational shorthand to make certain constructs more convenient. For
+example, consider the [`fancy-app`][1] package. This package provides an `#%app`[racket] macro that
+behaves just like normal function application *unless* one or more underscores are used. In that
+case, the function application is converted to a lambda that takes as input one argument for each
+underscore. For example, `(format "Hello ~a" _)` is equivalent to
+`(lambda (v) (format "Hello ~a" v))`. This is especially useful for whipping up quick lambdas as
+arguments to functions like `map` and `filter`:
+
+```racket
+> (require fancy-app)
+> (map (format "Hello ~a, how are you today?" _)
+       (list "Alice" "Bob" "Eve"))
+'("Hello Alice, how are you today?"
```

Comment:
Want to change this to "Hello Adele, can you hear me?" ?
(I guess, Adele Paul .... can't think of a third)